File: ROS2/ros2_ws/src/basics_pubsub/basics_pubsub/simple_sub.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class SimpleSub(Node):
    def __init__(self):
        super().__init__('simple_sub')
        self.sub = self.create_subscription(String, 'chatter', self.on_msg, 10)

    def on_msg(self, msg: String):
        self.get_logger().info(f"heard: {msg.data}")

def main():
    rclpy.init()
    node = SimpleSub()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("Shutting down")
    finally:
        node.destroy_node()
        # Only call shutdown if the context is still active
        if rclpy.ok():
            rclpy.shutdown()



if __name__ == '__main__':
    main()

File: ROS2/ros2_ws/src/basics_pubsub_cpp/src/simple_sub.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"



class SimpleSub : public rclcpp::Node {
public:
  SimpleSub() : Node("simple_sub_cpp") {
    sub_ = this->create_subscription<std_msgs::msg::String>(
      "chatter_cpp", 10,
      [this](std_msgs::msg::String::SharedPtr msg) {
        RCLCPP_INFO(this->get_logger(), "heard: %s", msg->data.c_str());
      }
    );
  }
private:
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
};

int main(int argc, char * argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleSub>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}